[RFR] Stash command

chrisngyn opened the pull request:
Fixes #286.

Instead of implementing stashing functionality, we'll be relying on [keep](https://github.com/OrkoHunter/keep).

Thinking of wrapping the functionality in a `-s` flag instead of having to type `echo -e 'while loop python\n$(howdoi while loop python)\n\n' | keep new | keep list` instead (how you would store it in a one liner; keep normally will prompt you for stdin two times).

Introduces new flags:
`-sn` `-se` `-sv`

Update -- will be removing edit functionality, and changing names to `-save` `-view` `-remove` and `-clear` (updated respectively in the README)

chrisngyn commented:
Ready for your review @gleitz!

![hi](https://user-images.githubusercontent.com/35744669/85628928-f867c800-b63e-11ea-89c6-365113914f13.png)

chrisngyn commented on setup.py:
```diff
@@ -63,6 +63,7 @@ def read(*names):
     entry_points={
         'console_scripts': [
             'howdoi = howdoi.howdoi:command_line_runner',
+            'keep = keep new | keep list'
```
Oops. In reviewing my own code I forgot that we weren't going with this route anymore.

kx-chen commented on howdoi/howdoi.py:
```diff
@@ -487,6 +508,14 @@ def command_line_runner():
         return
 
+    if args['stash_view']:
+        os.system('keep list')
```
Could we add some basic error handling here? Looks like `os.system` returns the exit status, but could be a bit more complex on other operating systems. https://docs.python.org/3/library/os.html#os.system

kx-chen commented on howdoi/howdoi.py:
```diff
@@ -425,6 +425,21 @@ def _get_cache_key(args):
+        # replace ' quotes with " to prevent interference with echo command quotes.
+        # replace \\n with \\ n to prevent echoing newline.
+        description = answer.replace("'", '"').replace('\n', '\\n').replace('\\n', '\\ n')
+        str_builder = "echo '" + cmd + "\n" + description + "\n\n' | keep new > /dev/null"
```
Will `/dev/null` work on Windows?

gleitz commented:
Thoughts from our call:
- Think about the names of the flags. `--save`? `-sn` vs `-ss` (stash save). I don't know yet but maybe a brainstorming would be fun.
- sr could give you a list of possible queries to delete, and let you choose one
- Remove the editing functionality and leave that up to keep?

chrisngyn commented on howdoi/howdoi.py:
```diff
@@ -432,6 +434,57 @@ def _get_cache_key(args):
     return str(args) + __version__
 
 
+def print_stash(sl = []):
```
Could we use a more descriptive name than `sl`?

chrisngyn commented on howdoi/howdoi.py:
```diff
@@ -432,6 +434,57 @@ def _get_cache_key(args):
+
+        for cmd, fields in commands.items():
+            stash_list.append(
+                '\033[4m\033[1m$ ' + fields['alias'] + '\033[0m\n\n' 
```
We should probably add some comments explaining what this is -- if someone sees this they'll definitely question what it means.

chrisngyn commented on howdoi/howdoi.py:
```diff
@@ -432,6 +434,57 @@ def _get_cache_key(args):
+    commands = keep_utils.read_commands()
+    if commands is not None and cmd_key in commands:
+        keep_utils.remove_command(cmd_key)
+        print('\n\033[1m\033[92m"' + title + '" removed from stash.\033[0m' + '\n\ncommand key data --- ' + cmd_key + '\n')
```
Add comments about how we format this

chrisngyn commented:
@gleitz ready for your review.

Since our last meeting;
- replaced `-sn` with `-save`
- replaced `-sv` with `-view`
- replaced `-sc` with `-clear`
- removed `-se` entirely (editing feature)
- added `-remove` functionality, users will be prompted from their stash which entry they want to delete

gleitz commented on howdoi/howdoi.py:
```diff
@@ -432,6 +434,61 @@ def _get_cache_key(args):
+    return str(stash_args)
+
+
+def remove_stash_command(cmd_key, title):
```
perhaps you can call these functions `_stash_remove`, `_stash_save`, `_stash_empty`, `_stash_list` which will better map to the names we use externally (more readable). Also please use the leading `_` to show that these are internal functions. Also please group all the stash functions together.

gleitz commented on howdoi/howdoi.py:
```diff
@@ -432,6 +434,61 @@ def _get_cache_key(args):
+        print('\n\033[1m\033[92m"' + title + '" removed from stash.\033[0m' + '\n\ncommand key data --- ' + cmd_key + '\n')
+    else:
+        # \033[1m\033[91m to bold and red text.
+        print('\n\033[1m\033[91m"' + title + '" not found in stash.\033[0m' + '\n\ncommand key data --- ' + cmd_key + '\n')
```
A good opportunity to use string formatting

gleitz commented on howdoi/howdoi.py:
```diff
@@ -432,6 +434,61 @@ def _get_cache_key(args):
+        for cmd, fields in commands.items():
+            # \033[4m\033[1m to bold and underline text.
+            stash_list.append(
+                '\033[4m\033[1m$ ' + fields['alias'] + '\033[0m\n\n' 
```
use format rather than string concatenation

gleitz commented on howdoi/howdoi.py:
```diff
@@ -472,13 +529,48 @@ def get_parser():
+
+    last_index = len(stash_list)
+    # \033[1m to bold text.
+    prompt = "\033[1m> Select a stash command to remove [1-" + str(last_index) + "] (0 to cancel): \033[0m"
```
remove string concatenation

chrisngyn commented on README.rst:
```diff
@@ -98,12 +98,24 @@ Usage
+      -save, --stash-save   stash a howdoi answer
+      -view, --stash-view   view your stash
+      -remove, --stash-remove remove an entry in your stash
+      -empty, --stash-empty empty your stash
```
I also thought it was weird, but `-v` was reserved for version, `-e` was reversed for engine, etc. and I didn't want half the stash functions to be one letter while the others were a full word. We can discuss this more before merging!

chrisngyn commented:
@gleitz ready for your review.

Here's a quick use case for how you would use the stashing feature if you wanted to try it yourself:

```
$ howdoi --save while loop python
$ howdoi --save format date bash
$ howdoi --view
$ howdoi --remove (will be prompted which to delete, we have the appropriate error handling too with bad inputs)
$ howdoi --empty (will be prompted to confirm)
```

Since our last meeting;

- used global variables for text formatting which we use `.format()` on
- replaced -save, -view, -remove, -empty with --save, --view, --remove, --empty
- updated the README accordingly

I know you asked us if we could use `f-strings` and we tried that, but we saw that it failed the Travis CI tests for Python versions 3.4 and 3.5 due to `f-strings` being adding in Python 3.6.

We *could* have done something like this;

```python
if version < 3.6:
    manual formatting...
else:
    f-string...
```

But this seemed redundant to me. I know that you wanted us to use `f-strings` to clean up the code base and make it easier to read, but due to Python 3.4 and 3.5 not having `f-strings`, we would have to use the manual formatting regardless and have an if statement to determine when we would use them.

My solution was to still go with your suggestion of having global variables, so now instead of seeing `"{}".format('\n\033m...')` you would see `"{}".format(BOLD)` instead. I think this is easier to read than how it was before.

Let us know what you think! 😄 🎉

gleitz commented on howdoi/howdoi.py:
```diff
@@ -432,6 +441,66 @@ def _get_cache_key(args):
+    description = fields['desc']
+    item_num = index + 1
+    if index == -1:
+        return '{}{}$ {}{}\n\n{}\n'.format(UNDERLINE, BOLD, title, END_FORMAT, description)
```
I think it would be more readable if you used the following:
```
'{underline}{bold}$ {title}{end_format}\n\n{description}\n'.format(underline=UNDERLINE,
                                                                   bold=BOLD,
                                                                   title=title,
                                                                   end_format=END_FORMAT,
                                                                   description=description)
```

gleitz commented on howdoi/howdoi.py:
```diff
@@ -432,6 +441,66 @@ def _get_cache_key(args):
+
+def _get_stash_key(args):
+    stash_args = {}
+    ignore_keys = ['save', 'view', 'remove', 'empty', 'tags']
```
Please make these globals

gleitz commented on howdoi/howdoi.py:
```diff
@@ -494,6 +598,23 @@ def command_line_runner():
             _print_err('Clearing cache failed')
         return
 
+    if args['view']:
```
use the globals here

gleitz commented on howdoi/howdoi.py:
```diff
@@ -432,6 +441,66 @@ def _get_cache_key(args):
+        stash_list = ['\nSTASH LIST:']
+        commands = keep_utils.read_commands()
+        if commands is None or len(commands.items()) == 0:
+            print('No commands found in stash. Add a command with "howdoi -save <query>".')
```
use the global STASH_SAVE here. Also it's `--save`

gleitz commented on howdoi/howdoi.py:
```diff
@@ -476,9 +545,44 @@ def get_parser():
+        print_stash(stash_list)
+
+    last_index = len(stash_list)
+    prompt = "{}> Select a stash command to remove [1-{}] (0 to cancel): {}".format(BOLD, last_index, END_FORMAT)
```
Use named string formatting here

gleitz commented on howdoi/howdoi.py:
```diff
@@ -476,9 +545,44 @@ def get_parser():
+        if user_input == 0:
+            return
+        elif user_input < 1 or user_input > last_index:
+            print("\n{}Input index is invalid.{}".format(RED, END_FORMAT))
```
Use named formatting here

gleitz commented on howdoi/howdoi.py:
```diff
@@ -476,9 +545,44 @@ def get_parser():
+        _stash_remove(cmd_key, cmd_name)
+        return
+    except ValueError:
+        print("\n{}Invalid input. Must specify index of command.{}".format(RED, END_FORMAT))
```
Use named formatting here

gleitz commented on howdoi/howdoi.py:
```diff
@@ -494,6 +598,23 @@ def command_line_runner():
+    if args['remove'] and len(args['query']) == 0:
+        commands = keep_utils.read_commands()
+        if commands is None or len(commands.items()) == 0:
+            print('No commands found in stash. Add a command with "howdoi --save <query>".')
```
Use the global STASH_SAVE here

gleitz commented:
I think we may need to issue a "keep init"? This is my first time using keep on a fresh install of howdoi:

```
» python -m howdoi.howdoi format date bash --save
Traceback (most recent call last):
  File "/Users/gleitz/.homebrew/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/gleitz/.homebrew/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/gleitz/projects/howdoi/howdoi/howdoi.py", line 670, in <module>
    command_line_runner()
  File "/Users/gleitz/projects/howdoi/howdoi/howdoi.py", line 659, in command_line_runner
    utf8_result = howdoi(args).encode('utf-8', 'ignore')
  File "/Users/gleitz/projects/howdoi/howdoi/howdoi.py", line 554, in howdoi
    return _parse_cmd(args, res)
  File "/Users/gleitz/projects/howdoi/howdoi/howdoi.py", line 520, in _parse_cmd
    _stash_save(cmd_key, title, answer)
  File "/Users/gleitz/projects/howdoi/howdoi/howdoi.py", line 511, in _stash_save
    keep_utils.save_command(cmd_key, answer, title)
  File "/Users/gleitz/.virtualenvs/howdoi/lib/python3.7/site-packages/keep/utils.py", line 192, in save_command
    with open(json_path, 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/gleitz/.keep/commands.json'
```

chrisngyn commented:
Will propose a fix for this today.
